Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp

```diff
@@ -0,0 +1,475 @@
+
+            if (should_move_position)
+                m_parent_container.value()->swap_widget_positions(this, relative_position().translated(m_current_mouse_position));
+            else
+                set_view_state(ViewState::Detached);
+        }
+
+        m_is_dragging = false;
+        set_override_cursor(Gfx::StandardCursor::Arrow);
+        update();
+    }
+}
+
+void DynamicWidgetContainer::mousemove_event(MouseEvent& event)
+{
+    if (m_hover_over_control_indicator != m_expand_collapse_button.united(m_detach_widget_button).contains(event.position())) {
+        m_hover_over_control_indicator = !m_hover_over_control_indicator;
+        update();
+    }
+    auto changed_cursor = m_is_dragging ? Gfx::StandardCursor::Move : Gfx::StandardCursor::Arrow;
+    if (m_move_widget_knurl.contains(event.position()) && !m_is_dragging)
+        changed_cursor = Gfx::StandardCursor::Hand;
+
+    if (event.buttons() == MouseButton::Primary && !m_is_dragging) {
+        m_is_dragging = true;
+        m_drag_start_location = event.position();
+        changed_cursor = Gfx::StandardCursor::Move;
+    }
+
+    if (m_is_dragging) {
+        m_current_mouse_position = event.position();
+        if (m_parent_container.has_value()) {
+            m_parent_container.value()->check_has_move_target(relative_position().translated(m_current_mouse_position), MoveTargetOp::SetTarget);
+        }
+        update();
+    }
+
+    if (changed_cursor != override_cursor())
+        set_override_cursor(changed_cursor);
+}
+
+void DynamicWidgetContainer::doubleclick_event(MouseEvent& event)
+{
+    if (event.button() != MouseButton::Primary || is_container_with_individual_order())
+        return;
+
+    if (view_state() == ViewState::Expanded)
+        set_view_state(ViewState::Collapsed);
+    else if (view_state() == ViewState::Collapsed)
+        set_view_state(ViewState::Expanded);
+}
+
+void DynamicWidgetContainer::leave_event(Core::Event&)
+{
+    m_hover_over_control_indicator = false;
+    update();
+}
+
+void DynamicWidgetContainer::resize_event(ResizeEvent&)
+{
+    constexpr int button_width_and_internal_margin = 24;
+    m_expand_collapse_button.set_x(width() - (button_width_and_internal_margin + content_margins().right()));
+    m_detach_widget_button.set_x(m_expand_collapse_button.x() - 18);
+
+    // Check if there is any content to display, and hide ourself if there would be nothing to display.
+    // This allows us to make the whole section not taking up space if child-widget visibility is maintained outside.
+    if (m_previous_frame_style.has_value() && height() != 0) {
+        m_controls_spacer->set_visible(true);
+        set_frame_style(m_previous_frame_style.value());
+        m_previous_frame_style = {};
+
+        deferred_invoke([&]() {
+            invalidate_layout();
+        });
```

Comment:
Pretty sure you can call invalidate_layout directly.

Reply:
Hm, it is not working as expected without the deferred invoke and resulted in the widgets not always being drawn completely. So I didn't changed the existing solution as that working as expected.